spec/mem/tree_borrows/state_machine.md

```diff
@@ -107,10 +133,25 @@ impl LocationState {
         &mut self,
         access_kind: AccessKind,
         node_relation: NodeRelation,
+        protected: Protected,
     ) -> Result {
-        self.permission = self.permission.transition(access_kind, node_relation)?;
+        self.permission = self.permission.transition(access_kind, node_relation, protected, self.accessed)?;
         self.accessed = self.accessed.transition(node_relation);
```

Comment:
Protected nodes cannot get disabled. So what I would expect is that there is a check here: if `permission` is now `Disabled` and `protected.yes()`, then we cause UB. Isn't that the entire point of protectors? How can things work when that is not done...?

Reply:
In this implementation, we cause UB when the permission **becomes** `Disabled`. This is slightly different from Miri, sorry for confusing

Reply:
Yes, it's implemented differently to miri but achieves the same effect. I also think it is somewhat more clean.

Reply:
Why do you think this is more clean? I find the Miri version more clean -- a single general principle, uniformly enforced, is better than having to copy the same logic in several places.

Reply:
I also really like the Miri version :)

But when I was implementing this, I thought that this version could explicitly show all the state transitions, directly corresponding to the state machine.

Reply:
To me, the state machine is what Miri has, and the rule that "when protected transitions to disabled, that's UB" is implemented outside the state machine.

Reply:
> the rule that "when protected transitions to disabled, that's UB" is implemented outside the state machine.

This makes sense to me. Fixed it :)